examples/step-67/step-67.cc

```diff
@@ -0,0 +1,2305 @@
+        }
+    }
+
+  private:
+    std::vector<double> bi;
+    std::vector<double> ai;
+  };
+
+
+
+  // @sect3{Implementation of point-wise operations of the Euler equations}
+
+  // In the following functions, we implement the various problem-specific
+  // operators pertaining to the Euler equations. Each function acts on the
+  // vector of conserved variables $[\rho, \rho\mathbf{u}, E]$ that we hold in
+  // the solution vectors, and computes various derived quantities.
+  //
+  // First out is the computation of the velocity, that we derive from the
+  // momentum variable $\rho \mathbf{u}$ by division by $\rho$. One thing to
+  // note here is that we decorate all those functions with the keyword
+  // `DEAL_II_ALWAYS_INLINE`. This is a special macro that maps to a
+  // compiler-specific keyword that tells the compiler to never create a
+  // function call for any of those functions, and instead move the
+  // implementation inline to where they are called. This is critical for
+  // performance because we repeatedly call into some of those functions: For
+  // example, we both use the velocity for the computation of the flux further
+  // down, but also for the computation of the pressure. Keeping these
+  // functions inline means that the repeated use is seen by the compiler
+  // during the optimization passes, and it eventually only keeps a single
+  // call around. If it were a separate function, it gets more complicated or
+  // impossible because already computed temporary information cannot be
+  // passed around.
+  //
+  // Another trick we apply is a separate variable for the inverse density
+  // $\frac{1}{\rho}$. This enables the compiler to only perform a single
+  // division for the flux, despite the division being used at several
+  // places. As divisions are around ten to twenty times as expensive as
+  // multiplications or additions, avoiding redundant divisions is crucial for
+  // performance. We note that taking the inverse first and later multiplying
+  // with it is not equivalent to a division in floating point arithmetic due
+  // to roundoff effects, so the compiler is not allowed to do it with
+  // standard optimization flags. However, it is also not particularly
+  // difficult to write the code in the right way.
+  //
+  // To summarize, the chosen strategy of always inlining and careful
+  // definition of expensive arithmetic operations allows us to write compact
+  // code without passing all intermediate results around, despite making sure
+  // that the code maps to excellent machine code.
+  template <int dim, typename Number>
+  inline DEAL_II_ALWAYS_INLINE //
+    Tensor<1, dim, Number>
+    euler_velocity(const Tensor<1, dim + 2, Number> &conserved_variables)
+  {
+    const Number           inverse_density = 1. / conserved_variables[0];
+    Tensor<1, dim, Number> velocity;
+    for (unsigned int d = 0; d < dim; ++d)
+      velocity[d] = conserved_variables[1 + d] * inverse_density;
```

Comment:
You mean like we do in `step-33`? I can do that, but it makes code pretty long because I would write `inverse_density = Number(1.)/conserved_variables[density_component]` and then `velocity[d] = conserved_variables[first_velocity_component + d] * inverse_density;`. This will make the program 20-40 lines longer I suspect (as I'm close to the 80 column limit already now), but if there is consensus to do that, I will.

Reply:
I think it would help to understand the code. You do not have to do for me: I am familiar with this ;)